gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionContext.java

```diff
@@ -215,6 +217,34 @@ public void removePendingSelectionSourceForOrderBy(String source) {
     pendingSelectionSourcesForOrderBy.remove(source);
   }
 
+  public void removeSelectionAttributes(String source, Set<String> attributes) {
+    if (!sourceToSelectionExpressionMap.containsKey(source)) {
+      return;
+    }
+
+    Predicate<Expression> retainExpressionPredicate =
+        expression ->
+            Sets.intersection(ExpressionReader.extractColumns(expression), attributes).isEmpty();
```

Comment:
nit: collecting all of `ExpressionReader.extractColumns(expression)` in a set and then taking the intersection with attributes would be more efficient I guess, as it would need a single pass over the attributes array

Reply:
That would be more trickier to maintain the fact, which expression lead to a set of corresponding attributes. Because, we need those expressions which lie outside of the intersection.

Basically a set difference, but with a custom comparator. I can take a look at it separately, if there is a cleaner way to do that